File: python/scripts/generate.py
#!/usr/bin/env python3
"""Codegen: emit ``src/smooth_operator/_generated.py`` from the language-neutral
JSON Schemas in ``../spec`` using datamodel-code-generator (pydantic v2).

Strategy
--------
Every schema file in the spec is self-contained: all ``$ref``\\ s point at internal
``#/$defs/...`` definitions (verified — no cross-file refs). The *interesting*
shapes live under ``$defs``:

  * ``envelope.schema.json`` — top-level ``oneOf`` of ``ActionEnvelope`` /
    ``EventEnvelope``; the real shapes (incl. ``ErrorObject``) are under ``$defs``.
  * ``actions/*.schema.json`` — top-level ``oneOf`` over ``$defs`` (``Request`` /
    ``Response`` + shared helpers like ``GeneralAgentResponse``,
    ``ConversationMessage``).
  * ``events/*.schema.json`` — a flat top-level object (the event frame itself).
  * ``domain/*.schema.json`` — a flat top-level object (the domain entity), some
    with their own ``$defs`` (``MessageContent`` / ``ContentItem``).

We merge everything into **one** synthetic JSON Schema whose ``$defs`` holds every
named model we want emitted, deduplicating shared defs by ``title`` (``ErrorObject``,
``ConversationMessage``, ``GeneralAgentResponse`` appear in more than one file).
datamodel-code-generator then emits one pydantic model per ``$def`` in a single
pass, so cross-references between them resolve cleanly.

Run via ``uv run python scripts/generate.py``.
"""

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
SPEC_DIR = HERE.parent.parent / "spec"
OUT_FILE = HERE.parent / "src" / "smooth_operator" / "_generated.py"

# Order matters only for readability of the merged doc; codegen sorts its output.
SUBDIRS = ["", "actions", "events", "domain"]


def list_schemas() -> list[Path]:
    out: list[Path] = []
    for sub in SUBDIRS:
        d = SPEC_DIR / sub if sub else SPEC_DIR
        for p in sorted(d.glob("*.schema.json")):
            out.append(p)
    return out


def title_for(name: str, schema: dict, file_title: str) -> str:
    """Pick a stable model name for a ``$def`` (its ``title`` if present)."""
    return str(schema.get("title") or f"{file_title}{name}")


def make_rewriter(rename: dict[str, str]):
    """Rewrite intra-file ``#/$defs/X`` refs to point at the merged top-level
    ``#/$defs/<Title>`` namespace.

    Every named ``$def`` is re-keyed under its model *title* in the merged doc, so a
    ref like ``#/$defs/Request`` (inside ``send-message``) must become
    ``#/$defs/SendMessageRequest`` (the title of that def).
    """

    def walk(node: object) -> object:
        if isinstance(node, dict):
            new: dict = {}
            for k, v in node.items():
                if k == "$ref" and isinstance(v, str) and v.startswith("#/$defs/"):
                    local = v[len("#/$defs/") :]
                    target = rename.get(local, local)
                    new[k] = f"#/$defs/{target}"
                else:
                    new[k] = walk(v)
            return new
        if isinstance(node, list):
            return [walk(x) for x in node]
        return node

    return walk


def build_merged_schema() -> dict:
    merged_defs: dict[str, dict] = {}
    seen: set[str] = set()

    for path in list_schemas():
        raw = json.loads(path.read_text())
        file_title = str(raw.get("title") or path.stem)

        local_defs = raw.get("$defs") or {}
        has_one_of = isinstance(raw.get("oneOf"), list)

        if has_one_of and local_defs:
            # Map each local $def name -> its global title, for ref rewriting.
            rename = {name: title_for(name, sub, file_title) for name, sub in local_defs.items()}
            rewrite = make_rewriter(rename)
            for name, sub in local_defs.items():
                title = rename[name]
                if title in seen:
                    continue  # shared def already emitted from an earlier file
                seen.add(title)
                model = {k: v for k, v in sub.items() if k != "$defs"}
                model = rewrite(model)  # fix internal refs
                model["title"] = title
                merged_defs[title] = model
        else:
            # Flat top-level object (events, domain). It may carry its own $defs.
            rename = {name: title_for(name, sub, file_title) for name, sub in local_defs.items()}
            rewrite = make_rewriter(rename)

            # Promote the file's own $defs into the merged namespace.
            for name, sub in local_defs.items():
                title = rename[name]
                if title not in seen:
                    seen.add(title)
                    model = {k: v for k, v in sub.items() if k != "$defs"}
                    model = rewrite(model)
                    model["title"] = title
                    merged_defs[title] = model

            title = file_title
            if title in seen:
                continue
            seen.add(title)
            top = {k: v for k, v in raw.items() if k not in ("$defs", "$schema", "$id")}
            top = rewrite(top)
            top["title"] = title
            merged_defs[title] = top

    return {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "$id": "https://smooth-agent.dev/spec/_merged.schema.json",
        "title": "SmoothAgentProtocol",
        "type": "object",
        "$defs": merged_defs,
    }


HEADER = '''"""AUTO-GENERATED — do not edit by hand.

Generated from the JSON Schemas in ../spec by scripts/generate.py
Run ``uv run python scripts/generate.py`` to regenerate after a schema change.

These are faithful pydantic v2 reflections of the wire schemas: snake_case Python
attributes with camelCase aliases (``populate_by_name = True``), so attribute access
is idiomatic while ``model_dump(by_alias=True)`` round-trips the camelCase wire form.
"""
# ruff: noqa
# fmt: off
'''


def main() -> int:
    merged = build_merged_schema()

    tmp = OUT_FILE.parent / "_merged.schema.json"
    OUT_FILE.parent.mkdir(parents=True, exist_ok=True)
    tmp.write_text(json.dumps(merged, indent=2))

    raw_out = OUT_FILE.parent / "_generated_raw.py"
    cmd = [
        sys.executable,
        "-m",
        "datamodel_code_generator",
        "--input",
        str(tmp),
        "--input-file-type",
        "jsonschema",
        "--output",
        str(raw_out),
        "--output-model-type",
        "pydantic_v2.BaseModel",
        "--target-python-version",
        "3.11",
        "--use-annotated",
        "--use-field-description",
        "--snake-case-field",  # snake_case attrs
        "--use-default-kwarg",
        "--reuse-model",
        "--use-standard-collections",
        "--use-union-operator",
        "--collapse-root-models",
        "--disable-timestamp",
        "--allow-population-by-field-name",  # populate_by_name = True
    ]
    print("Running:", " ".join(cmd[1:]))
    subprocess.run(cmd, check=True)

    body = raw_out.read_text()
    # datamodel-code-generator writes its own header; replace it with ours.
    if body.startswith('"""'):
        # drop everything up to and including the first generated header block
        idx = body.find("from __future__")
        if idx == -1:
            idx = body.find("\n\n")
        body = body[idx:]
    OUT_FILE.write_text(HEADER + "\n" + body.lstrip())

    raw_out.unlink(missing_ok=True)
    tmp.unlink(missing_ok=True)
    print(f"Wrote {OUT_FILE}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/scripts/generate.ts
/**
 * Codegen: emit `src/generated/types.ts` from the language-neutral JSON Schemas
 * in `../spec`.
 *
 * Strategy
 * --------
 * Every schema file in the spec is self-contained: all `$ref`s point at internal
 * `#/$defs/...` definitions (verified — no cross-file refs). So we can run
 * `json-schema-to-typescript` per file and concatenate the results into a single
 * `types.ts`, deduplicating shared interface names that several files happen to
 * declare (e.g. `ErrorObject`).
 *
 * The envelope file is special: it is a top-level `oneOf` of `ActionEnvelope` and
 * `EventEnvelope` with the interesting shapes living under `$defs`. We compile its
 * `$defs` so consumers get `ActionEnvelope` / `EventEnvelope` / `ErrorObject`.
 *
 * The action/event/domain files likewise hang their real shapes off `$defs`
 * (Request/Response) or are a flat top-level object (events). We feed each file's
 * effective schema to json-schema-to-typescript with a stable title so the emitted
 * interface name is predictable.
 */
import { readFile, writeFile, readdir } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import { dirname, join, relative } from 'node:path';
import { compile, type JSONSchema } from 'json-schema-to-typescript';

const __dirname = dirname(fileURLToPath(import.meta.url));
const SPEC_DIR = join(__dirname, '..', '..', 'spec');
const OUT_FILE = join(__dirname, '..', 'src', 'generated', 'types.ts');

const COMPILE_OPTS = {
    additionalProperties: false,
    bannerComment: '',
    style: { singleQuote: true, tabWidth: 4, trailingComma: 'all' as const },
    declareExternallyReferenced: true,
    enableConstEnums: false,
};

async function listSchemas(): Promise<string[]> {
    const out: string[] = [];
    for (const sub of ['', 'actions', 'events', 'domain', 'interactions']) {
        const dir = sub ? join(SPEC_DIR, sub) : SPEC_DIR;
        const entries = await readdir(dir, { withFileTypes: true });
        for (const e of entries) {
            if (e.isFile() && e.name.endsWith('.schema.json')) {
                out.push(join(dir, e.name));
            }
        }
    }
    return out.sort();
}

/**
 * For a file whose top-level is a `oneOf` over `$defs` (envelope, actions), strip
 * the `oneOf`/`$ref` wrapper and return one synthetic schema per named `$def` so
 * each gets its own exported interface. Flat object files are returned as-is.
 */
function expand(schema: JSONSchema, fileTitle: string): JSONSchema[] {
    const defs = (schema.$defs ?? {}) as Record<string, JSONSchema>;
    const hasOneOf = Array.isArray(schema.oneOf);

    if (hasOneOf && Object.keys(defs).length > 0) {
        // Emit each named $def as a standalone schema, carrying $defs along so
        // internal refs still resolve.
        return Object.entries(defs).map(([name, def]) => ({
            ...def,
            $defs: defs,
            title: def.title ?? `${fileTitle}${name}`,
        }));
    }

    // Flat top-level object (the event schemas, domain schemas).
    return [{ ...schema, title: schema.title ?? fileTitle }];
}

async function main(): Promise<void> {
    const files = await listSchemas();
    const seenNames = new Set<string>();
    const blocks: string[] = [];

    for (const file of files) {
        const raw = JSON.parse(await readFile(file, 'utf8')) as JSONSchema;
        const fileTitle = String(raw.title ?? '');
        const subSchemas = expand(raw, fileTitle);

        for (const sub of subSchemas) {
            // json-schema-to-typescript keys the top-level interface name off `title`.
            const ts = await compile(sub, String(sub.title ?? fileTitle), COMPILE_OPTS);
            // Drop interfaces whose name we've already emitted from an earlier file
            // (shared defs like ErrorObject appear in multiple files).
            const filtered = stripOrphanComments(dropDuplicateDeclarations(ts, seenNames));
            if (filtered.trim().length > 0) {
                blocks.push(`// ── from ${relative(SPEC_DIR, file)} ──\n${filtered.trim()}\n`);
            }
        }
    }

    const header = `/**
 * AUTO-GENERATED — do not edit by hand.
 *
 * Generated from the JSON Schemas in ../spec by scripts/generate.ts
 * Run \`pnpm generate\` to regenerate after a schema change.
 */
/* eslint-disable */

`;

    await writeFile(OUT_FILE, header + blocks.join('\n'), 'utf8');
    console.log(`Wrote ${OUT_FILE} (${blocks.length} declaration blocks from ${files.length} schema files).`);
}

/**
 * json-schema-to-typescript emits each interface/type as a top-level declaration.
 * Walk the emitted source, drop any declaration whose name we have already seen,
 * and register newly-seen names. This deduplicates shared `$defs` (e.g.
 * `ErrorObject`, `ConversationMessage`) that several files compile independently.
 */
function dropDuplicateDeclarations(source: string, seen: Set<string>): string {
    const declRe = /^export (?:interface|type) (\w+)\b/;
    const lines = source.split('\n');
    const kept: string[] = [];

    let skipping = false;
    let depth = 0;
    let currentIsType = false;

    for (const line of lines) {
        const m = line.match(declRe);
        if (m && depth === 0) {
            const name = m[1]!;
            currentIsType = line.startsWith('export type');
            if (seen.has(name)) {
                skipping = true;
            } else {
                seen.add(name);
                skipping = false;
            }
        }

        if (!skipping) kept.push(line);

        // Track brace depth so we know when an interface block ends.
        for (const ch of line) {
            if (ch === '{') depth++;
            else if (ch === '}') depth = Math.max(0, depth - 1);
        }

        // `export type X = ...;` declarations are single-line (depth stays 0) —
        // stop skipping once the statement terminates.
        if (currentIsType && depth === 0 && line.includes(';')) {
            skipping = false;
            currentIsType = false;
        } else if (!currentIsType && depth === 0 && skipping && line.includes('}')) {
            skipping = false;
        }
    }

    return kept.join('\n');
}

/**
 * After dropping a duplicate declaration we may be left with its leading JSDoc
 * block dangling with no declaration after it. Remove any `/** ... *​/` comment
 * that is not immediately followed (ignoring blank lines) by a declaration.
 */
function stripOrphanComments(source: string): string {
    return source.replace(/\/\*\*[\s\S]*?\*\/\s*(?=\n\s*(?:\/\*\*|$))/g, '').replace(/\n{3,}/g, '\n\n');
}

main().catch((err) => {
    console.error(err);
    process.exit(1);
});
